servicetalk-http-api/src/main/java/io/servicetalk/http/api/OffloadingFilter.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.api;
+
+import io.servicetalk.concurrent.api.Executor;
+import io.servicetalk.concurrent.api.Single;
+
+import static java.util.function.Function.identity;
+
+/**
+ * An {@link StreamingHttpServiceFilterFactory} implementation which offloads filters using a provided strategy.
+ */
+final class OffloadingFilter implements StreamingHttpServiceFilterFactory {
+
+    private final HttpExecutionStrategy strategy;
+    private final StreamingHttpServiceFilterFactory offloaded;
+
+    /**
+     * @param strategy Execution strategy for the offloaded filters
+     * @param offloaded Filters to be offloaded
+     */
+    OffloadingFilter(HttpExecutionStrategy strategy, StreamingHttpServiceFilterFactory offloaded) {
+        this.strategy = strategy;
+        this.offloaded = offloaded;
+    }
+
+    @Override
+    public StreamingHttpServiceFilter create(StreamingHttpService service) {
+        return new StreamingHttpServiceFilter(offloaded.create(service)) {
+
+            @Override
+            public Single<StreamingHttpResponse> handle(final HttpServiceContext ctx,
+                                                        StreamingHttpRequest request,
+                                                        final StreamingHttpResponseFactory responseFactory) {
+                Executor e = null != strategy.executor() ? strategy.executor() : ctx.executionContext().executor();
```

Comment:
`strategy.executor()` in theory may change in-between checking against null and then saving the reference. cache the reference in a local variable to avoid this.

Reply:
Currently there are no strategies with non-final executors and with the soon-to-follow #1695 it will be moot.